File: BOJ/01000/1213.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "../input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const s = input[0];
const cnt = Array.from({ length: 26 }).fill(0);
let ret = "";
let mid = "";
let fail = 0;

for (let i = 0; i < s.length; i++) {
  cnt[s[i].charCodeAt() - "A".charCodeAt()]++;
}

for (let i = "Z".charCodeAt(); i >= "A".charCodeAt(); i--) {
  const idx = i - "A".charCodeAt();

  if (cnt[idx]) {
    if (cnt[idx] % 2 === 1) {
      fail++;
      mid = String.fromCharCode(i);
      cnt[idx]--;
    }
  }

  if (fail === 2) break;

  for (let j = 0; j < cnt[idx]; j += 2) {
    ret = String.fromCharCode(i) + ret + String.fromCharCode(i);
  }
}

if (mid) {
  let midIdx = Math.floor(ret.length / 2);
  ret = ret.slice(0, midIdx) + mid + ret.slice(midIdx);
}

if (fail == 2) {
  console.log("I'm Sorry Hansoo");
} else {
  console.log(ret);
}

File: BOJ/01000/1213.cpp
#include <bits/stdc++.h>
using namespace std;

string s, ret;
int cnt[200], fail;
char mid;
int main() {
  cin >> s;
  for (char it : s) cnt[it]++;

  for (int i = 'Z'; i >= 'A'; i--) {
    if (cnt[i]) {
      if (cnt[i] % 2 == 1) {
        fail++;
        mid = char(i);
        // 중앙에 넣을 알파벳 1개 소모
        cnt[i]--;
      }

      if (fail == 2) break;

      // 알파벳을 2개씩 소모
      for (int j = 0; j < cnt[i]; j += 2) {
        ret = char(i) + ret;
        ret += char(i);
      }
    }
  }

  if (mid) {
    ret.insert(ret.begin() + ret.size() / 2, mid);
  }

  if (fail == 2) {
    cout << "I'm Sorry Hansoo\n";
  } else {
    cout << ret << "\n";
  }

  return 0;
}

// cin >> s;
// for(char a : s)cnt[a]++;
// for(int i = 'Z'; i >= 'A'; i--){
// 	if(cnt[i]){
// 		if(cnt[i] & 1){
// 			mid = char(i);flag++;
// 			cnt[i]--;
// 		}
// 		if(flag == 2)break;
// 		for(int j = 0; j < cnt[i]; j += 2){
// 			ret = char(i) + ret;
// 			ret += char(i);
// 		}
// 	}
// }
// if(mid)ret.insert(ret.begin() + ret.size() / 2, mid);
// if(flag == 2)cout << "I'm Sorry Hansoo\n";
// else cout << ret << "\n";

// 시간초과
// string v, s;
// bool found;
// int main() {
//   cin >> s;

//   for (int i = 0; i < s.size(); i++) {
//     v.push_back(s[i]);
//   }

//   sort(v.begin(), v.end());

//   do {
//     string a, b;

//     for (auto it : v) a.push_back(it);

//     b.resize(a.size());
//     reverse_copy(a.begin(), a.end(), b.begin());

//     if (a == b) {
//       found = true;
//       cout << a;
//       break;
//     }
//   } while (next_permutation(v.begin(), v.end()));

//   if (!found) {
//     cout << "I'm Sorry Hansoo";
//   }

//   return 0;
// }